File: 백준/Gold/27211. 도넛 행성/도넛 행성.py
from collections import deque

def bfs(y, x):
    q = deque()
    q.append((y, x))

    while q:
        y, x = q.popleft()
        for dy, dx in ((0, -1), (-1, 0), (0, 1), (1, 0)):
            ny, nx = y+dy, x+dx
            ny, nx = (ny+n)%n, (nx+m)%m
            if visited[ny][nx]==0 and arr[ny][nx]==0:
                visited[ny][nx]=1
                q.append((ny, nx))


n, m = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(n)]
visited = [[0]*m for _ in range(n)]

cnt = 0
for i in range(n):
    for j in range(m):
        if arr[i][j]==0 and visited[i][j]==0:
            bfs(i, j)
            cnt += 1

print(cnt)

File: 백준/Gold/27211. 도넛 행성/도넛 행성.java
import java.util.*;
import java.io.*;

public class Main {

    static int n, m;
    static int[][] arr;
    static boolean[][] visited;
    static class Pair {
        int y, x;

        public Pair(int y, int x) {
            this.y = y;
            this.x = x;
        }
    }
    static int[] dy = {0, 1, 0, -1};
    static int[] dx = {1, 0, -1, 0};
    static int count = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        arr = new int[n][m];
        for (int i = 0; i<n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j<m; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        visited = new boolean[n][m];
        for (int i = 0; i<n; i++) {
            for (int j = 0; j<m; j++) {
                if (arr[i][j]==0 && !visited[i][j]) {
                    count++;
                    bfs(i, j);
                }
            }
        }
        System.out.println(count);
    }

    public static void bfs(int i, int j) {
        Queue<Pair> q = new LinkedList<>();
        q.add(new Pair(i, j));
        visited[i][j] = true;

        while (!q.isEmpty()) {
            Pair curr = q.poll();
            int y = curr.y, x = curr.x;

            for (int d = 0; d<4; d++) {
                int ny = y+dy[d];
                int nx = x+dx[d];
                if (ny<0) ny = n-1;
                if (nx<0) nx = m-1;
                if (ny>=n) ny = 0;
                if (nx>=m) nx = 0;
                if (ny<0 || ny>=n || nx<0 || nx>=m) continue;
                if (arr[ny][nx]==0 && !visited[ny][nx]) {
                    visited[ny][nx] = true;
                    q.add(new Pair(ny, nx));
                }
            }
        }
    }
}